docs/adr/0008-allowing-more-types-as-errors.md

```diff
@@ -0,0 +1,139 @@
+# Allowing more types to be used as errors.
+
+* Status: proposed
+* Deciders: Hopefully consensus on #1613
+
+## Current UniFFI error handling
+
+UniFFI functions currently must report error returns via an `Enum` - ie, the `E` in `Result<T, E>` must be an Enum.
+
+This ADR proposes moving towards a model where any type could be an `E`, and specfically, proposes allowing any uniffi "interface"
+
+Assuming this isn't controversial, it will also make decisions about how such errors should be represented in the foreign bindings we maintain.
+
+### The current situation
+When an `Enum` is used as an error, on the foreign side it's not an `Enum` at all - it's a distinct type -
+it derives from the local `Exception` class and isn't usable as the actual `Enum` behind it.
+
+### The options for interfaces.
+Consider the UDL
+```idl
+interface ErrorDetails { string message(); };
+[Throws=ErrorDetails] void do_it();
+``````
+and Rust
+```rust
+struct ErrorDetails {}
+impl ErrorDetails {
+    fn message(&self) -> String { todo!() }
+}
+fn do_it() -> Result<(), ErrorDetails> { Err(ErrorDetails {}) }
+```
+
+How should this be captured on the foreign side?
+
+### A) Create an "error only" type
+
+This is like the status quo re enums. It means:
+* There's no "non error" version of the interface.
+* A struct simply appearing as an error changes how the foreign side sees that struct significantly.
+* The type can't be used exactly as it could in non-error contexts.
```

Comment:
I don't think this is a requirement.  It seems possible to allow these types to be returned, inputted as arguments, etc.  It also seems possible to let the foreign code call into their API outside of a `catch` block.

I think the real issue is the bullet above.  It would be very weird for the foreign code to have a type that can be used both as a regular object and an exception.